wandb/sdk/wandb_settings.py

```diff
@@ -1116,6 +1121,99 @@ def _path_convert(*args: str) -> str:
         """Join path and apply os.path.expanduser to it."""
         return os.path.expanduser(os.path.join(*args))
 
+    @staticmethod
+    def _unpack_slurm_node_list(regex_string: str) -> Tuple[str, ...]:
+        """Unpacks a regex with the SLURM node list string into a list of strings."""
+        pattern = re.compile(r"\[([\d,-]+)\]")
+        match = pattern.search(regex_string)
+
+        if match is None:
+            return (regex_string,)
+
+        segments = match.group(1).split(",")
+        unpacked_strings = []
+
+        for segment in segments:
+            if "-" in segment:
+                start_number, end_number = (int(x) for x in segment.split("-"))
+                for number in range(start_number, end_number + 1):
+                    unpacked_strings.append(
+                        regex_string[: match.start()]
+                        + str(number)
+                        + regex_string[match.end() :]
+                    )
+            else:
+                unpacked_strings.append(
+                    regex_string[: match.start()]
+                    + segment
+                    + regex_string[match.end() :]
+                )
+
+        return tuple(unpacked_strings)
+
+    def _auto_open_metrics_endpoints(
+        self, value: Optional[Mapping[str, str]]
+    ) -> Optional[Mapping[str, str]]:
+        if value is not None:
+            # the user has explicitly set metrics endpoint(s), don't try to auto-discover
+            return value
+
+        # discover DCGM Exporter metrics endpoints
+        if not self._stats_auto_dcgm_exporter:
+            # auto-discovery not enabled
+            return None
+
+        slurm_node_id = os.environ.get("SLURM_NODEID")
+        slurm_nodelist = os.environ.get("SLURM_NODELIST")
+        slurm_nodename = os.environ.get("SLURMD_NODENAME")
+
+        # capture all metrics from the rank 0 node
+        if slurm_node_id != "0" or slurm_nodelist is None or slurm_nodename is None:
+            return None
+
+        # drop this node:
+        node_list = [
+            node
+            for node in self._unpack_slurm_node_list(slurm_nodelist)
+            if node != slurm_nodename
+        ]
+
+        return {
+            **{
+                "node1": "http://localhost:9400/metrics",
```

Comment:
this is a bold assumption; port 9400 is used by the DCGM exporter by default, but is configurable. but if it's not there, OpenMetrics.is_available() will return False, so nothing criminal will happen.